gemini/function-calling/parallel_function_calling.ipynb

```diff
@@ -0,0 +1,858 @@
+    "\n",
+    "Gemini is a family of generative AI models developed by Google DeepMind that is designed for multimodal use cases. The Gemini API gives you access to the Gemini Pro and Gemini Pro Vision models.\n",
+    "\n",
+    "[Function Calling](https://cloud.google.com/vertex-ai/docs/generative-ai/multimodal/function-calling) in Gemini lets you create a description of a function in your code, then pass that description to a language model in a request. The response from the model includes the name of a function that matches the description and the arguments to call it with.\n",
+    "\n",
+    "In this tutorial, you'll learn how to work with parallel function calling within Gemini Function Calling, including:\n",
+    "    \n",
+    "- Handling parallel function calls for repeated functions\n",
+    "- Working with parallel function calls across multiple functions\n",
+    "- Extracting multiple function calls from a Gemini response\n",
+    "- Calling multiple functions and returning them to Gemini\n",
+    "\n",
+    "### What is parallel function calling?\n",
+    "\n",
+    "In previous versions of the Gemini Pro models (prior to May 2024), Gemini would return two or more chained function calls if the model determined that more than one function call was needed before returning a natural language summary. Here, a chained function call means that you get the first function call response, return the API data to Gemini, get a second function call response, return the API data to Gemini, and so on.\n",
+    "\n",
+    "In recent versions of specific Gemini Pro models (from May 2024 and on), Gemini has the ability to return two or more function calls in parallel (i.e., two or more function call responses within the first function call response object). Parallel function calling allows you to fan out and parallelize your API calls or other actions that you perform in your application code, so you don't have to work through each function call response and return one-by-one! Refer to the [Gemini Function Calling documentation](https://cloud.google.com/vertex-ai/generative-ai/docs/multimodal/function-calling) for more information on which Gemini model versions support parallel function calling.\n",
+    "\n",
+    "\n",
+    "<img src=\"https://storage.googleapis.com/github-repo/generative-ai/gemini/function-calling/parallel-function-calling-in-gemini.png\" width=\"80%\">"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "r11Gu7qNgx1p"
+   },
+   "source": [
+    "## Getting Started\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "No17Cw5hgx12"
+   },
+   "source": [
+    "### Install Vertex AI SDK and other required packages"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {
+    "id": "tFy3H3aPgx12",
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "!pip3 install --upgrade --user --quiet google-cloud-aiplatform google-play-scraper"
```

Comment:
Is this google-play-scraper library official? Or violate any TOS? Since this is an official google notebook, just want to be safe. It is a good example

Reply:
Good point, I (incorrectly) assumed there was an equivalent Google-managed API for this. Let me dig around and see if I can find an API or client lib to substitute in here that would be equally as simple.

Reply:
It's probably not a big deal, it's just accessing public information.

Reply:
I'm reworking this notebook to use the OSS `wikipedia` Python lib since Wikipedia has a public API for this and the library is working within the parameters of their ToS. It's a good point and is always the hardest part of these samples, as you know. So I appreciate you mentioning this!

Reply:
Done, I've reworked this notebook to use the `wikipedia` Python package and make calls to Wikipedia API.